### PR TITLE
Power and network modulisation

### PR DESCRIPTION
**Pull request by baz2142**
## Description of changes

**[Packages] add package SshKeysHelper**
Ssh keys-related helpers

**[Modules] Host: add power control option**

**[Targets] Lenovo-X1-Carbon: refactoring**
Moving all needed configurations into
corresponding modules and replacing hardcoded
ssh-keys stuff

**[Targets] Lenovo-X1-Carbon: enabling Power Control**

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

## Testing

Test power off and reboot is working on Lenovo X1, nothing is broken

### CODE REVIEW

**Comment by baz2142** [comment]
@leivos-unikie, can you test it?

**Comment by baz2142** [comment]
@leivos-unikie, can you test again on X1 as I made changes. And also any another platform without Power Control feature -- just to be sure everything is intact

**Comment by leivos-unikie** [comment]
**Tested on Lenovo-X1**
- nix build github:baz2142/ghaf/power-and-network-modulisation#lenovo-x1-carbon-gen11-debug-installer
- Reboot and power-off buttons work
- All apps can be launched
- ci-test-automation test-suite pass, performance results ok


**Tested on Orin AGX**
- nix build github:baz2142/ghaf/power-and-network-modulisation#nvidia-jetson-orin-agx-debug-flash-script
- apps launch (except Element)
- net-vm does not boot

```
[0;ghaf@ghaf-host: ~ghaf@ghaf-host:~]$ microvm -l
net-vm: current(nixos-system-net-vm-23.11pre-git), not booted: systemctl start microvm@net-vm.service

[0;ghaf@ghaf-host: ~ghaf@ghaf-host:~]$ sudo systemctl start microvm@net-vm.service
[sudo] password for ghaf: 

[0;ghaf@ghaf-host: ~microvm -l
net-vm: current(nixos-system-net-vm-23.11pre-git), not booted: systemctl start microvm@net-vm.service

```

**Comment by leivos-unikie** [comment]
Checked `journalctl | grep net-vm` on Orin AGX. Terminal flooded with this:

Jul 28 09:30:47 ghaf-host microvm@net-vm[2637]: qemu-system-aarch64: -device virtio-9p-pci,fsdev=fs1,mount_tag=waypipe-ssh-public-key: cannot initialize fsdev 'fs1': failed to open '/run/waypipe-ssh-public-key': No suc
h file or directory
Jul 28 09:30:47 ghaf-host systemd[1]: microvm@net-vm.service: Main process exited, code=exited, status=1/FAILURE
Jul 28 09:30:47 ghaf-host systemd[1]: microvm@net-vm.service: Failed with result 'exit-code'.
Jul 28 09:30:48 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier
Jul 28 09:30:48 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:52 ghaf-host systemd[1]: microvm@net-vm.service: Scheduled restart job, restart counter is at 24.
Jul 28 09:30:52 ghaf-host systemd[1]: Stopped MicroVM 'net-vm'.
Jul 28 09:30:52 ghaf-host systemd[1]: Setup MicroVM 'net-vm' MACVTAP interfaces was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/macvtap-interfaces).
Jul 28 09:30:52 ghaf-host systemd[1]: microvm-pci-devices@net-vm.service: Deactivated successfully.
Jul 28 09:30:52 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' devices for passthrough.
Jul 28 09:30:52 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' devices for passthrough...
Jul 28 09:30:52 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' devices for passthrough...
Jul 28 09:30:52 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' TAP interfaces...
Jul 28 09:30:52 ghaf-host systemd[1]: VirtioFS daemons for MicroVM 'net-vm' was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/virtiofs).
Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: /sys/bus/pci/devices/0001:01:00.0 /var/lib/microvms/net-vm
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:52 ghaf-host kernel: device tap-net-vm left promiscuous mode
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: Making VFIO device 50 accessible for user
Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: /var/lib/microvms/net-vm
Jul 28 09:30:52 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' devices for passthrough.
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed to start LLDP client: No such device
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Could not process link message: No such device
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Link DOWN
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier
Jul 28 09:30:52 ghaf-host systemd[1]: microvm-tap-interfaces@net-vm.service: Deactivated successfully.
Jul 28 09:30:52 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' TAP interfaces.
Jul 28 09:30:52 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' TAP interfaces...
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Link UP
Jul 28 09:30:52 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' TAP interfaces.
Jul 28 09:30:52 ghaf-host systemd[1]: Starting MicroVM 'net-vm'...
Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Configuring with /etc/systemd/network/11-netvm.network.
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:52 ghaf-host kernel: device tap-net-vm entered promiscuous mode
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state
Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state
Jul 28 09:30:53 ghaf-host systemd[1]: Started MicroVM 'net-vm'.
Jul 28 09:30:53 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier
Jul 28 09:30:53 ghaf-host microvm@net-vm[2701]: qemu-system-aarch64: -device virtio-9p-pci,fsdev=fs1,mount_tag=waypipe-ssh-public-key: cannot initialize fsdev 'fs1': failed to open '/run/waypipe-ssh-public-key': No suc
h file or directory
Jul 28 09:30:53 ghaf-host systemd[1]: microvm@net-vm.service: Main process exited, code=exited, status=1/FAILURE
Jul 28 09:30:53 ghaf-host systemd[1]: microvm@net-vm.service: Failed with result 'exit-code'.
Jul 28 09:30:53 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier
Jul 28 09:30:53 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:58 ghaf-host systemd[1]: microvm@net-vm.service: Scheduled restart job, restart counter is at 25.
Jul 28 09:30:58 ghaf-host systemd[1]: Stopped MicroVM 'net-vm'.
Jul 28 09:30:58 ghaf-host systemd[1]: Setup MicroVM 'net-vm' MACVTAP interfaces was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/macvtap-interfaces).
Jul 28 09:30:58 ghaf-host systemd[1]: microvm-pci-devices@net-vm.service: Deactivated successfully.
Jul 28 09:30:58 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' devices for passthrough.
Jul 28 09:30:58 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' devices for passthrough...
Jul 28 09:30:58 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' devices for passthrough...
Jul 28 09:30:58 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' TAP interfaces...
Jul 28 09:30:58 ghaf-host systemd[1]: VirtioFS daemons for MicroVM 'net-vm' was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/virtiofs).
Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: /sys/bus/pci/devices/0001:01:00.0 /var/lib/microvms/net-vm
Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier
Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state
Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state
Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:58 ghaf-host kernel: device tap-net-vm left promiscuous mode
Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state
Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: Making VFIO device 50 accessible for user
Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: /var/lib/microvms/net-vm
Jul 28 09:30:58 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' devices for passthrough.
Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed to start LLDP client: No such device
Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Could not process link message: No such device
Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed
Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Link DOWN
Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier

**Comment by baz2142** [comment]
> Checked `journalctl | grep net-vm` on Orin AGX. Terminal flooded with this:
> 
> Jul 28 09:30:47 ghaf-host microvm@net-vm[2637]: qemu-system-aarch64: -device virtio-9p-pci,fsdev=fs1,mount_tag=waypipe-ssh-public-key: cannot initialize fsdev 'fs1': failed to open '/run/waypipe-ssh-public-key': No suc h file or directory Jul 28 09:30:47 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Main process exited, code=exited, status=1/FAILURE Jul 28 09:30:47 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Failed with result 'exit-code'. Jul 28 09:30:48 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier Jul 28 09:30:48 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Scheduled restart job, restart counter is at 24. Jul 28 09:30:52 ghaf-host systemd[1]: Stopped MicroVM 'net-vm'. Jul 28 09:30:52 ghaf-host systemd[1]: Setup MicroVM 'net-vm' MACVTAP interfaces was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/macvtap-interfaces). Jul 28 09:30:52 ghaf-host systemd[1]: [microvm-pci-devices@net-vm.service](mailto:microvm-pci-devices@net-vm.service): Deactivated successfully. Jul 28 09:30:52 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:52 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:52 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:52 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' TAP interfaces... Jul 28 09:30:52 ghaf-host systemd[1]: VirtioFS daemons for MicroVM 'net-vm' was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/virtiofs). Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: /sys/bus/pci/devices/0001:01:00.0 /var/lib/microvms/net-vm Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host kernel: device tap-net-vm left promiscuous mode Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: Making VFIO device 50 accessible for user Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: /var/lib/microvms/net-vm Jul 28 09:30:52 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed to start LLDP client: No such device Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Could not process link message: No such device Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Link DOWN Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier Jul 28 09:30:52 ghaf-host systemd[1]: [microvm-tap-interfaces@net-vm.service](mailto:microvm-tap-interfaces@net-vm.service): Deactivated successfully. Jul 28 09:30:52 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' TAP interfaces. Jul 28 09:30:52 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' TAP interfaces... Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Link UP Jul 28 09:30:52 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' TAP interfaces. Jul 28 09:30:52 ghaf-host systemd[1]: Starting MicroVM 'net-vm'... Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Configuring with /etc/systemd/network/11-netvm.network. Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host kernel: device tap-net-vm entered promiscuous mode Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state Jul 28 09:30:53 ghaf-host systemd[1]: Started MicroVM 'net-vm'. Jul 28 09:30:53 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier Jul 28 09:30:53 ghaf-host microvm@net-vm[2701]: qemu-system-aarch64: -device virtio-9p-pci,fsdev=fs1,mount_tag=waypipe-ssh-public-key: cannot initialize fsdev 'fs1': failed to open '/run/waypipe-ssh-public-key': No suc h file or directory Jul 28 09:30:53 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Main process exited, code=exited, status=1/FAILURE Jul 28 09:30:53 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Failed with result 'exit-code'. Jul 28 09:30:53 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier Jul 28 09:30:53 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:58 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Scheduled restart job, restart counter is at 25. Jul 28 09:30:58 ghaf-host systemd[1]: Stopped MicroVM 'net-vm'. Jul 28 09:30:58 ghaf-host systemd[1]: Setup MicroVM 'net-vm' MACVTAP interfaces was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/macvtap-interfaces). Jul 28 09:30:58 ghaf-host systemd[1]: [microvm-pci-devices@net-vm.service](mailto:microvm-pci-devices@net-vm.service): Deactivated successfully. Jul 28 09:30:58 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:58 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:58 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:58 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' TAP interfaces... Jul 28 09:30:58 ghaf-host systemd[1]: VirtioFS daemons for MicroVM 'net-vm' was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/virtiofs). Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: /sys/bus/pci/devices/0001:01:00.0 /var/lib/microvms/net-vm Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:58 ghaf-host kernel: device tap-net-vm left promiscuous mode Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: Making VFIO device 50 accessible for user Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: /var/lib/microvms/net-vm Jul 28 09:30:58 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed to start LLDP client: No such device Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Could not process link message: No such device Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Link DOWN Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-

> Checked `journalctl | grep net-vm` on Orin AGX. Terminal flooded with this:
> 
> Jul 28 09:30:47 ghaf-host microvm@net-vm[2637]: qemu-system-aarch64: -device virtio-9p-pci,fsdev=fs1,mount_tag=waypipe-ssh-public-key: cannot initialize fsdev 'fs1': failed to open '/run/waypipe-ssh-public-key': No suc h file or directory Jul 28 09:30:47 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Main process exited, code=exited, status=1/FAILURE Jul 28 09:30:47 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Failed with result 'exit-code'. Jul 28 09:30:48 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier Jul 28 09:30:48 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Scheduled restart job, restart counter is at 24. Jul 28 09:30:52 ghaf-host systemd[1]: Stopped MicroVM 'net-vm'. Jul 28 09:30:52 ghaf-host systemd[1]: Setup MicroVM 'net-vm' MACVTAP interfaces was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/macvtap-interfaces). Jul 28 09:30:52 ghaf-host systemd[1]: [microvm-pci-devices@net-vm.service](mailto:microvm-pci-devices@net-vm.service): Deactivated successfully. Jul 28 09:30:52 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:52 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:52 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:52 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' TAP interfaces... Jul 28 09:30:52 ghaf-host systemd[1]: VirtioFS daemons for MicroVM 'net-vm' was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/virtiofs). Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: /sys/bus/pci/devices/0001:01:00.0 /var/lib/microvms/net-vm Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host kernel: device tap-net-vm left promiscuous mode Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: Making VFIO device 50 accessible for user Jul 28 09:30:52 ghaf-host microvm-pci-devices@net-vm[2675]: /var/lib/microvms/net-vm Jul 28 09:30:52 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed to start LLDP client: No such device Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Could not process link message: No such device Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Link DOWN Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier Jul 28 09:30:52 ghaf-host systemd[1]: [microvm-tap-interfaces@net-vm.service](mailto:microvm-tap-interfaces@net-vm.service): Deactivated successfully. Jul 28 09:30:52 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' TAP interfaces. Jul 28 09:30:52 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' TAP interfaces... Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Link UP Jul 28 09:30:52 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' TAP interfaces. Jul 28 09:30:52 ghaf-host systemd[1]: Starting MicroVM 'net-vm'... Jul 28 09:30:52 ghaf-host systemd-networkd[1132]: tap-net-vm: Configuring with /etc/systemd/network/11-netvm.network. Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:52 ghaf-host kernel: device tap-net-vm entered promiscuous mode Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:52 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state Jul 28 09:30:53 ghaf-host systemd[1]: Started MicroVM 'net-vm'. Jul 28 09:30:53 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier Jul 28 09:30:53 ghaf-host microvm@net-vm[2701]: qemu-system-aarch64: -device virtio-9p-pci,fsdev=fs1,mount_tag=waypipe-ssh-public-key: cannot initialize fsdev 'fs1': failed to open '/run/waypipe-ssh-public-key': No suc h file or directory Jul 28 09:30:53 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Main process exited, code=exited, status=1/FAILURE Jul 28 09:30:53 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Failed with result 'exit-code'. Jul 28 09:30:53 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier Jul 28 09:30:53 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:58 ghaf-host systemd[1]: [microvm@net-vm.service](mailto:microvm@net-vm.service): Scheduled restart job, restart counter is at 25. Jul 28 09:30:58 ghaf-host systemd[1]: Stopped MicroVM 'net-vm'. Jul 28 09:30:58 ghaf-host systemd[1]: Setup MicroVM 'net-vm' MACVTAP interfaces was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/macvtap-interfaces). Jul 28 09:30:58 ghaf-host systemd[1]: [microvm-pci-devices@net-vm.service](mailto:microvm-pci-devices@net-vm.service): Deactivated successfully. Jul 28 09:30:58 ghaf-host systemd[1]: Stopped Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:58 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:58 ghaf-host systemd[1]: Starting Setup MicroVM 'net-vm' devices for passthrough... Jul 28 09:30:58 ghaf-host systemd[1]: Stopping Setup MicroVM 'net-vm' TAP interfaces... Jul 28 09:30:58 ghaf-host systemd[1]: VirtioFS daemons for MicroVM 'net-vm' was skipped because of an unmet condition check (ConditionPathExists=/var/lib/microvms/net-vm/current/share/microvm/virtiofs). Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: /sys/bus/pci/devices/0001:01:00.0 /var/lib/microvms/net-vm Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Gained carrier Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered blocking state Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered forwarding state Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:58 ghaf-host kernel: device tap-net-vm left promiscuous mode Jul 28 09:30:58 ghaf-host kernel: virbr0: port 1(tap-net-vm) entered disabled state Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: Making VFIO device 50 accessible for user Jul 28 09:30:58 ghaf-host microvm-pci-devices@net-vm[2717]: /var/lib/microvms/net-vm Jul 28 09:30:58 ghaf-host systemd[1]: Finished Setup MicroVM 'net-vm' devices for passthrough. Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed to start LLDP client: No such device Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Could not process link message: No such device Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Failed Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Link DOWN Jul 28 09:30:58 ghaf-host systemd-networkd[1132]: tap-net-vm: Lost carrier

Thank you, it should help me to fix it

**Comment by milva-unikie** [comment]
Net-vm is still not booting on AGX. We will do more comprehensive testing after this issue is resolved.

**Comment by baz2142** [comment]
@milva-unikie, it should work now on both of the platforms. Can you check?

**Comment by milva-unikie** [comment]
Tested on Lenovo-X1
- reboot and power-off buttons work
- apps launch from icons
- all of the vm:s have started
- test-automation passes
- performance good

Tested on Orin-AGX & Orin-AGX Cross & Orin-NX
- apps launch from icons
- net-vm has started
- test-automation passes

File: modules/microvm/power-control.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+in
+  with lib; {
+    options.ghaf.host.powercontrol.enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Enable host power control";
+    };
+
+    config = mkIf cfg.enable {
+      ghaf.host.networking.enable = true;
```
Comment (brianmcgillion) [inline modules/microvm/power-control.nix:18]: why setting this under power control it is already set in each target/definition

File: modules/microvm/virtualization/microvm/netvm.nix [hunk 2]
```diff
@@ -9,6 +9,14 @@
   configHost = config;
   vmName = "net-vm";
   macAddress = "02:00:00:01:01:01";
+
+  isGuiVmEnabled = config.ghaf.virtualization.microvm.guivm.enable;
```
Comment (brianmcgillion) [inline modules/microvm/virtualization/microvm/netvm.nix:13]: why is there a dependency on the ui vm?

File: modules/microvm/power-control.nix (hunk 1 above)
Comment (baz2142) [reply modules/microvm/power-control.nix:18]: Power control cannot work without network, so I enable it as a depedency. Maybe here some other ways to require network or indicate it needs the network?

File: modules/microvm/virtualization/microvm/netvm.nix (hunk 2 above)
Comment (baz2142) [reply modules/microvm/virtualization/microvm/netvm.nix:13]: For example, AGX Orig doesn´t have GuiVM -- so we don´t have "waypipe" ssh-keys -- we don´t need to mount the keys to NetVM from the host and configure ssh client

Comment (baz2142) [reply modules/microvm/virtualization/microvm/netvm.nix:13]: As long as NetVM uses ssh keys from GuiVM -- it is dependent on GuiVM

File: modules/microvm/power-control.nix (hunk 1 above)
Comment (baz2142) [reply modules/microvm/power-control.nix]: Done

**Review by brianmcgillion — APPROVED** [review]

**Pull request merged**